### PR TITLE
Manon

### PR DESCRIPTION
**Pull request by manonreau**
I adapted the code to handle classification problems (including number of classes > 2): 
- task dependent loss function 
- softmax activation function on the network's output for classification problems
- accuracy calculation
- added an output_shape argument to the model (need to modify the diff networks input accordingly )

### CODE REVIEW

**Comment by manonreau** [comment]
updated foutnet.py ginet.py to provide output_shape as an input argument (optional, default = 1)

File: graphprot/NeuralNet.py [hunk 1]
```diff
@@ -53,61 +55,142 @@ def __init__(self, database, Net,
...
+            self.output_shape = len(self.classes)
+            try :
+                self.model = Net(dataset.get(0).num_features, self.output_shape).to(self.device)
+            except :
+                raise ValueError(
+                    f"The loaded model does not accept output_shape = {self.output_shape} argument \n\t"
+                    f"Check your input or adapt the model\n\t"
+                    f"Example :\n\t"
+                    f"def __init__(self, input_shape): --> def __init__(self, input_shape, output_shape) \n\t"
+                    f"self.fc2 = torch.nn.Linear(64, 1) --> self.fc2 = torch.nn.Linear(64, output_shape) \n\t")
```
Comment (NicoRenaud) [inline graphprot/NeuralNet.py:81]: can you add an else clause that would raise some error ?

File: graphprot/NeuralNet.py [hunk 2]
```diff
@@ -53,61 +55,142 @@ def __init__(self, database, Net,
...
+
             else:
                 t = time() - t0
-                print('Epoch [%04d] : train loss %e | time %1.2e sec.' % (
+                if acc is not None :
+                    print('Epoch [%04d] : train loss %e | accuracy %1.4e | time %1.2e sec. |' % (
+                    epoch, loss, acc, t))
+                else :
+                    print('Epoch [%04d] : train loss %e | time %1.2e sec. |' % (
                     epoch, loss, t))
```
Comment (NicoRenaud) [inline graphprot/NeuralNet.py:124]: I think it would be more elegan to have : 
```python
acc, loss = self._epoch()
print('Epoch [%04d] : train loss | accuracy .... ')
if validate:
    valid_acc, valid_train = self.eval()
    print('Epoch [%04d] : valid loss valid acc')
```
cause now you have a lot of code duplication which is never ideal :)

File: graphprot/NeuralNet.py [hunk 3]
```diff
@@ -53,61 +55,142 @@ def __init__(self, database, Net,
...
 
+
+    def calcAccuracy(self, prediction, target, reduce=True):
+
+        overlap = (prediction.argmax(dim=1)==target).sum()
```
Comment (NicoRenaud) [inline graphprot/NeuralNet.py:129]: nice oneliner ! Maybe explain what it does in the docstring (that we should start adding to all functions/classes). And I would call that variable accuracy then :)

File: graphprot/NeuralNet.py [hunk 4]
```diff
@@ -53,61 +55,142 @@ def __init__(self, database, Net,
...
+
+            if self.task == 'class':
+                out = F.softmax(out, dim=1)
+                data.y = torch.tensor([self.classes_idx[int(x)] for x in data.y])
+                acc.append(self.calcAccuracy(out, data.y))
+
+            else :
+                out = out.reshape(-1)
+                acc = None
+
```
Comment (NicoRenaud) [inline graphprot/NeuralNet.py:182]: you could encapsulate that in a function so that its's not duplicated in epoch and eval

**Review by NicoRenaud — APPROVED** [review]
Great ! I've made a few comments. If you can address them and fix them if you think it's worth it that great ! Then you can merge to master !!

**Comment by manonreau** [comment]
adapted the code to handle classification problems (including number of classes > 2)
